components/connector/internal/subject/subject.go

```diff
@@ -0,0 +1,123 @@
+	mappings  []subjectConsumerTypeMapping
+	ouPattern string
+}
+
+// NewProcessor returns a new subject processor configured with the given subject-to-consumer mapping, and subject organization unit pattern.
+// If the subject-to-consumer mapping is invalid, an error is returned.
+func NewProcessor(subjectConsumerTypeMappingConfig string, ouPattern string) (*processor, error) {
+	mappings, err := unmarshalMappings(subjectConsumerTypeMappingConfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while configuring subject processor")
+	}
+	for _, m := range mappings {
+		if err := m.validate(); err != nil {
+			return nil, err
+		}
+	}
+	return &processor{
+		mappings:  mappings,
+		ouPattern: ouPattern,
+	}, nil
+}
+
+// AuthIDFromSubjectFunc returns a function able to extract the authentication ID from a given certificate subject.
+func (p *processor) AuthIDFromSubjectFunc() func(subject string) string {
+	authIDFromMappingFunc := p.authIDFromMappings()
+	authIDFromOUsFunc := cert.GetRemainingOrganizationalUnit(p.ouPattern)
+	return func(subject string) string {
+		if authIDFromMapping := authIDFromMappingFunc(subject); authIDFromMapping != "" {
+			return authIDFromMapping
+		}
+		return authIDFromOUsFunc(subject)
+	}
+}
+
+// AuthSessionExtraFromSubjectFunc returns a function which returns consumer - specific auth session extra body
+// in case the subject matches any of the configured consumers from the mapping.
+func (p *processor) AuthSessionExtraFromSubjectFunc() func(subject string) map[string]interface{} {
+	return func(subject string) map[string]interface{} {
+		log.D().Infof("trying to extract auth session extra from subject %s", subject)
+		for _, m := range p.mappings {
+			log.D().Infof("trying to match subject pattern %s", m.Subject)
+			if subjectsMatch(subject, m.Subject) {
+				log.D().Infof("pattern matched subject!")
+				return cert.GetExtra(m.ConsumerType, m.TenantAccessLevel, m.InternalConsumerID)
+			}
+		}
+
+		return nil
+	}
+}
+
+func (p *processor) authIDFromMappings() func(subject string) string {
+	return func(subject string) string {
+		for _, m := range p.mappings {
+			if subjectsMatch(subject, m.Subject) {
+				return m.InternalConsumerID
+			}
+		}
+		return ""
+	}
+}
+
+func unmarshalMappings(mappingsConfig string) ([]subjectConsumerTypeMapping, error) {
+	var mappings []subjectConsumerTypeMapping
+	if err := json.Unmarshal([]byte(mappingsConfig), &mappings); err != nil {
+		return nil, errors.Wrap(err, "while unmarshalling mappings")
+	}
+
+	return mappings, nil
+}
+
+func subjectsMatch(actualSubject, expectedSubject string) bool {
+	actualSubjectComponents := strings.Split(actualSubject, ",")
+	expectedSubjectComponents := strings.Split(expectedSubject, ",")
+
+	if len(actualSubjectComponents) != len(expectedSubjectComponents) {
+		return false
+	}
+
+	for _, expectedSubjectComponent := range expectedSubjectComponents {
+		if !strings.Contains(actualSubject, strings.TrimSpace(expectedSubjectComponent)) {
```

Comment:
This flow is not covered by the unit tests but it seems like quite an important one because we don't want to match unexpected subjects. I'd add a unit test if it's not much effort for some reason or is covered by the e2e tests.